packages/ui/src/components/ResourceLoader/index.tsx

```diff
@@ -0,0 +1,85 @@
+import { useState, useEffect } from "react";
+
+import {
+  RemoteData,
+  remoteDataSuccess,
+  remoteDataFailure,
+  remoteDataLoading,
+  remoteDataNotAsked,
+  Kind,
+} from "../../utils/remote-data";
+
+interface ResourceLoaderProps<T> {
+  resource: Promise<T>;
+  renderLoading?: () => React.ReactElement | null;
+  renderFailure?: (err: Error) => React.ReactElement | null;
+  renderData?: (data: T) => React.ReactElement | null;
+  onSuccess?: (data: T) => void;
+  onError?: (err: Error) => void;
+}
+
+export function ResourceLoader<T>(props: ResourceLoaderProps<T>): React.ReactElement | null {
```

Comment:
Have you considered writing it entirely as a hook? What's the difference you think it would be.

Reply:
Making it a hook improves the flexibility. I would image the hook returns `{ error?: Error; data?: T; loading: boolean}`. Then the render part could be left to the user. Actually, there is [useSWR](https://www.vaultproject.io/) already does more than that.

Reply:
One benefit of the `ResourceLoader` is it has the default renders for `loading` and `error`, which saves me some time.